File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_admin_core/lib/src/services/provider_account_service.dart
import 'package:sdkwork_rtc_backend_sdk/sdkwork_rtc_backend_sdk.dart'
    as generated;

import '../admin_sdk_mapper.dart';
import '../backend_rtc_client.dart';
import '../models/paginated_list_result.dart';
import '../models/provider_account.dart';

class ProviderAccountService {
  final SdkworkBackendClient _client;

  ProviderAccountService(this._client);

  Future<PaginatedListResult<ProviderAccount>> list({
    String? provider,
    String? status,
    int? page,
    int? pageSize,
    String? cursor,
    String? search,
    String? sort,
  }) async {
    final response = await _client.rtcProviderAccounts.list(
      page,
      pageSize,
      cursor,
      sort,
      search,
    );
    return PaginatedListResult(
      items: backendResponseItems(response)
          .map(ProviderAccount.fromJson)
          .toList(),
      nextCursor: backendResponseNextCursor(response),
    );
  }

  Future<ProviderAccount?> get(String id) async {
    final response = await _client.rtcProviderAccounts.retrieve(id);
    final data = backendResponseData(response);
    return data == null ? null : ProviderAccount.fromJson(data);
  }

  Future<ProviderAccount?> create(ProviderAccountCommand command) async {
    final response = await _client.rtcProviderAccounts.create(
      generated.RtcProviderAccountCommand.fromJson(command.toJson()),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderAccount.fromJson(data);
  }

  Future<ProviderAccount?> update(String id, ProviderAccountCommand command) async {
    final response = await _client.rtcProviderAccounts.update(
      id,
      generated.RtcProviderAccountCommand.fromJson(command.toJson()),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderAccount.fromJson(data);
  }

  Future<ProviderAccount?> disable(String id, {String? reason}) async {
    final response = await _client.rtcProviderAccounts.disable(
      id,
      generated.RtcProviderAccountDisableRequest(reason: reason),
    );
    final data = backendResponseData(response);
    return data == null ? null : ProviderAccount.fromJson(data);
  }
}

File: apps/sdkwork-rtc-pc/packages/sdkwork-rtc-pc-admin-core/src/services/providerAccountService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type { ProviderAccount, ProviderAccountCommand } from "../types/providerAccount";
import { readSdkWorkItem, readSdkWorkListPage } from "../sdk/index.js";
import { resolveBackendRtcClient, type RtcBackendClientOptions, type RtcBackendClientSource } from "./backendClient";

interface ListResponse {
  items: ProviderAccount[];
  nextCursor?: string | null;
}

export class ProviderAccountService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async list(params?: {
    provider?: string;
    status?: string;
    page?: number;
    limit?: number;
    cursor?: string;
    search?: string;
    sort?: string;
  }): Promise<ListResponse> {
    const response = await this.client.rtcProviderAccounts.rtc.providerAccounts.list({
      page: params?.page,
      pageSize: params?.limit,
      cursor: params?.cursor,
      q: params?.search,
      sort: params?.sort,
    });
    const page = readSdkWorkListPage<ProviderAccount>(response);
    return {
      items: page.items,
      nextCursor: page.nextCursor ?? null,
    };
  }

  async get(id: string): Promise<ProviderAccount> {
    const response = await this.client.rtcProviderAccounts.rtc.providerAccounts.retrieve(id);
    if (!response) {
      throw new Error(`RTC provider account not found: ${id}`);
    }
    return readSdkWorkItem<ProviderAccount>(response);
  }

  async create(command: ProviderAccountCommand): Promise<ProviderAccount> {
    const response = await this.client.rtcProviderAccounts.rtc.providerAccounts.create(
      command as Parameters<
        typeof this.client.rtcProviderAccounts.rtc.providerAccounts.create
      >[0],
    );
    if (!response) {
      throw new Error("Invalid response: missing provider account data");
    }
    return readSdkWorkItem<ProviderAccount>(response);
  }

  async update(id: string, command: ProviderAccountCommand): Promise<ProviderAccount> {
    const response = await this.client.rtcProviderAccounts.rtc.providerAccounts.update(
      id,
      command as Parameters<
        typeof this.client.rtcProviderAccounts.rtc.providerAccounts.update
      >[1],
    );
    if (!response) {
      throw new Error("Invalid response: missing provider account data");
    }
    return readSdkWorkItem<ProviderAccount>(response);
  }

  async disable(id: string, reason?: string): Promise<ProviderAccount> {
    const response = await this.client.rtcProviderAccounts.rtc.providerAccounts.disable(id, {
      reason,
    });
    if (!response) {
      throw new Error("Invalid response: missing provider account data");
    }
    return readSdkWorkItem<ProviderAccount>(response);
  }
}
